### PR TITLE
Fix re-compilation bugs

### PR DESCRIPTION
**Pull request by mattdangerw**
The primary bug this is trying to fix is to get rid of weird re-compilation behaviors with our default compilation. E.g. create a `GemmaCausalLM`, generate some text without specifying a sampler, `compile()` again without a sampler, generation will have switched from `"greedy"` -> `"top_k"`. Create a `BertClassifier`, `fit()`, `compile()` again without specifying an optimizer, optimizer will have switch from `"adam"` to `"rmsprop"`.

The way I am trying to fix this is by leaning a little more heavily on the `"auto"` style option we introduced for `jit_compile`. KerasNLP tasks will by default use `loss="auto"` and `optimizer="auto"`, which resolve to a default for a given task.

Since we override compile with these in the signature, recompilation will not silently change behavior.

### CODE REVIEW

**Comment by mattdangerw** [comment]
talked with Francois offline, we should switch `sampler` to use "top_k" or "greedy" defaults.

**Review by fchollet — APPROVED** [review]

File: keras_nlp/models/causal_lm.py [hunk 1]
```diff
@@ -68,23 +68,79 @@ class CausalLM(Task):
...
+                applied for the token classification `CausalLM` task. See
+                `keras.Model.compile` and `keras.losses` for more info on
+                possible `loss` values.
+            weighted_metrics: `"auto"`, or a list of metrics to be evaluated by
+                the model during training and testing. Defaults to `"auto"`,
+                where a `keras.metrics.SparseCategoricalAccuracy` will be
+                applied to track the accuracy of the model at guessing masked
+                token values. See `keras.Model.compile` and `keras.metrics` for
+                more info on possible `weighted_metrics` values.
+            sampler: `"auto"`, a sampler name, or a `keras_nlp.samplers.Sampler`
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/models/causal_lm.py]: I think this was on a older version of the diff. This is fixed!
